Mathlib/Analysis/SpecialFunctions/BinEntropy.lean

```diff
@@ -0,0 +1,434 @@
+    · by_cases p_is_one : p = 1
+      · right; assumption
+      · have : 0 < binaryEntropy p := by
+          apply binaryEntropy_pos (Ne.lt_of_le ?_ domun)
+          refine Ne.lt_of_le ?_ ?_
+          repeat assumption
+          exact Iff.mp ne_comm pz
+        simp_all only [lt_self_iff_false]
+  · rw [binaryEntropy_eq']
+    cases h <;> simp [*]
+
+lemma zero_lt_log_two : 0 < log 2 := by
+  exact (log_pos_iff zero_lt_two).mpr one_lt_two
+
+/-- For probability `p < 0.5`, `binaryEntropy p < log 2`. -/
+lemma binaryEntropy_lt_log2_of_lt_half {p : ℝ} (pge0 : 0 ≤ p) (plehalf : p < 1/2) :
+    binaryEntropy p < log 2 := by
+  -- Proof by concavity of log.
+  rw [binaryEntropy_eq']
+  rw [show -p * p.log = -(p * p.log) by ring]
+  by_cases pz : p = 0
+  · simp [*]; exact zero_lt_log_two
+  · have invppos : 0 < 1/p := by positivity
+    have : 0 < 1 - p := by norm_num; linarith -- used implicitly by tactics.
+    have sub1pinvpos : 0 < 1 / (1 - p) := by positivity
+    have logConcave := (strictConcaveOn_log_Ioi.right
+      (x := 1/p) (y := 1/(1-p))) (a := p) (b := 1-p)
+      invppos sub1pinvpos (by norm_num; linarith) (by positivity)
+      (by norm_num; linarith) (by norm_num)
+    have : p • (1 / p) + (1 - p) • (1 / (1 - p)) = 2 := by field_simp; norm_num
+    rw [this] at logConcave
+    have := calc -(p * log p) - (1 - p) * log (1 - p)
+          _ = p * (-log p) + (1 - p) * (-log (1 - p)) := by ring
+          _ = p * log (1/p) + (1 - p) * log (1 / (1 - p)) := by rw [← log_inv]; norm_num
+    rw [this]
+    exact logConcave
+
+lemma binaryEntropy_lt_log2_of_gt_half {p : ℝ} : 1/2 < p → p ≤ 1 → binaryEntropy p < log 2 := by
+  intros
+  rw [← binaryEntropy_one_sub]
+  exact binaryEntropy_lt_log2_of_lt_half (by linarith) (by linarith)
+
+lemma binaryEntropy_eq_log2_iff_eq_half {p : ℝ} (pge0 : 0 ≤ p) (ple1 : p ≤ 1) :
+    binaryEntropy p = log 2 ↔ p = 1/2 := by
+  constructor <;> intro h
+  · by_cases h' : p < 1/2
+    · linarith [binaryEntropy_lt_log2_of_lt_half pge0 h']
+    · by_cases pgthalf : 1/2 < p
+      · have := binaryEntropy_lt_log2_of_gt_half pgthalf ple1
+        linarith
+      · linarith
+  · simp only [one_div, binaryEntropy_onehalf, h]
+
+lemma binaryEntropy_le_log2 {p : ℝ} (pge0 : 0 ≤ p) (ple1 : p ≤ 1) :
+    binaryEntropy p ≤ log 2 := by
+  by_cases hh: p = 1/2
+  · simp only [one_div, binaryEntropy_onehalf, le_refl, hh]
+  · by_cases gg: binaryEntropy p = log 2
+    · simp only [le_refl, gg]
+    · by_cases hhh: p < 1/2
+      · linarith [binaryEntropy_lt_log2_of_lt_half pge0 hhh]
+      · have : 1/2 < p := by
+          refine Ne.lt_of_le ?_ ?_
+          · simp only [not_lt] at hhh
+            intro a
+            simp_all only [not_true_eq_false]
+          · simp_all only [one_div, not_lt]
+        have := binaryEntropy_lt_log2_of_gt_half this ple1
+        exact LT.lt.le this
+
+/- The q-ary entropy function is continuous everywhere.
+This is due to definition of `Real.log` for negative numbers. -/
```

Comment:
```suggestion
/-- The q-ary entropy function is continuous everywhere.
This is due to definition of `Real.log` for negative numbers. -/
```
This isn't just extreme pickiness, there's actually a syntactic difference here: `/-- this -/` is the docstring of the lemma that follows, it's displayed in the documentation and in mouseovers in VSCode, whereas `/- this -/` is just a normal comment.

Reply:
Thanks. Fixed all docstrings.